aproxy.go

```diff
@@ -0,0 +1,404 @@
+	if response.StatusCode != 200 {
+		return nil, fmt.Errorf("proxy return %d response for connect request", response.StatusCode)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to receive http connect response from proxy: %w", err)
+	}
+	return conn, nil
+}
+
+func GetOriginalDst(conn *net.TCPConn) (*net.TCPAddr, error) {
+	file, err := conn.File()
+	defer func(file *os.File) {
+		err := file.Close()
+		if err != nil {
+			slog.Error("failed to close the duplicated TCP socket file descriptor")
+		}
+	}(file)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert connection to file: %w", err)
+	}
+	var sockaddr [16]byte
+	size := 16
+	_, _, e := syscall.Syscall6(
+		syscall.SYS_GETSOCKOPT,
+		file.Fd(),
+		syscall.SOL_IP,
+		80, // SO_ORIGINAL_DST
+		uintptr(unsafe.Pointer(&sockaddr)),
+		uintptr(unsafe.Pointer(&size)),
+		0,
+	)
+	if e != 0 {
+		return nil, fmt.Errorf("getsockopt SO_ORIGINAL_DST failed: errno %d", e)
+	}
+	return &net.TCPAddr{
+		IP:   sockaddr[4:8],
+		Port: int(binary.BigEndian.Uint16(sockaddr[2:4])),
+	}, nil
+}
+
+func RelayTcp(conn io.ReadWriter, proxyConn io.ReadWriteCloser, logger *slog.Logger) {
+	var closed atomic.Bool
+	go func() {
+		_, err := io.Copy(proxyConn, conn)
+		if err != nil && !closed.Load() {
+			logger.Error("failed to relay network traffic to proxy", "error", err)
+		}
+		closed.Store(true)
+		_ = proxyConn.Close()
+	}()
+	_, err := io.Copy(conn, proxyConn)
+	if err != nil && !closed.Load() {
+		logger.Error("failed to relay network traffic from proxy", "error", err)
+	}
+	closed.Store(true)
+}
+
+func RelayHttp(conn io.ReadWriter, proxyConn io.ReadWriteCloser, logger *slog.Logger) {
+	defer func() {
+		_ = proxyConn.Close()
+	}()
+	req, err := http.ReadRequest(bufio.NewReader(conn))
+	if err != nil {
+		logger.Error("failed to read HTTP request from connection", "error", err)
+		return
+	}
+	req.URL.Host = req.Host
+	req.URL.Scheme = "http"
+	req.Header.Set("Connection", "close")
+	if err := req.WriteProxy(proxyConn); err != nil {
+		logger.Error("failed to send HTTP request to proxy", "error", err)
+		return
+	}
+	resp, err := http.ReadResponse(bufio.NewReader(proxyConn), req)
+	if err != nil {
+		logger.Error("failed to read HTTP response from proxy", "error", err)
+		return
+	}
+	resp.Header.Set("Connection", "close")
+	if err := resp.Write(conn); err != nil {
+		logger.Error("failed to send HTTP response to connection", "error", err)
+		return
+	}
+}
+
+func HandleConn(conn net.Conn, proxy string) {
+	defer func() { _ = conn.Close() }()
+	logger := slog.With("src", conn.RemoteAddr())
+	dst, err := GetOriginalDst(conn.(*net.TCPConn))
+	if err != nil {
+		slog.Error("failed to get connection original destination", "error", err)
+		return
+	}
+	logger = logger.With("original_dst", dst)
+	var host string
+	var relay func(conn io.ReadWriter, proxyConn io.ReadWriteCloser, logger *slog.Logger)
+	var dialProxy func(proxy string) (net.Conn, error)
```

Comment:
To avoid these func variables with fairly long types, it might be simpler just to call `RelayTcp` and `DialProxyConnect` directly in the 443 switch case (and similar in the 80 case). It would get rid of these indirections and make the code a bit easier to follow.

Reply:
Good idea, updated, thanks!